Client/Frontend/Browser/New Tab Page/NewTabPageViewController.swift

```diff
@@ -424,8 +459,171 @@ class NewTabPageViewController: UIViewController, Themeable {
+                inNewTab: inNewTab,
+                switchingToPrivateMode: switchingToPrivateMode
+            )
+        case .itemAction(.toggledSource, let context):
+            let isEnabled = feedDataSource.isSourceEnabled(context.item.source)
+            feedDataSource.toggleSource(context.item.source, enabled: !isEnabled)
+            if isEnabled {
+                let alert = FeedActionAlertView(
+                    image: UIImage(imageLiteralResourceName: "disable.feed.source.alert"),
+                    title: Strings.BraveToday.disabledAlertTitle,
+                    message: String(format: Strings.BraveToday.disabledAlertBody, context.item.source.name)
+                )
+                alert.present(on: self)
+            }
+        case .itemAction(.longPressed(let context), _):
+            let alertController = UIAlertController(
+                title: context.title,
+                message: context.message,
+                preferredStyle: .actionSheet
+            )
+            for action in context.actions {
+                alertController.addAction(action)
+            }
+            present(alertController, animated: true)
+        }
+    }
+    
+    private var newContentAvailableDismissTimer: Timer? {
+        didSet {
+            oldValue?.invalidate()
+        }
+    }
+    
+    private func handleFeedStateChange(
+        _ oldValue: FeedDataSource.State,
+        _ newValue: FeedDataSource.State
+    ) {
+        guard let section = layout.braveTodaySection else { return }
+        
+        switch (oldValue, newValue) {
+        case (.loading, .loading):
+            // Nothing to do
+            break
+        case (.failure(let error1 as NSError),
+              .failure(let error2 as NSError)) where error1 == error2:
+            // Nothing to do
+            break
+        case (.loading(.failure(let error1 as NSError)),
+              .failure(let error2 as NSError)) where error1 == error2:
+            if let cell = collectionView.cellForItem(at: IndexPath(item: 0, section: section)) as? FeedCardCell<BraveTodayErrorView> {
+                cell.content.refreshButton.isLoading = false
+            }
+        case (_, .loading):
+            if collectionView.contentOffset.y == collectionView.contentInset.top ||
+                collectionView.numberOfItems(inSection: section) == 0 {
+                feedOverlayView.loaderView.isHidden = false
+                feedOverlayView.loaderView.start()
+                
+                if let section = layout.braveTodaySection {
+                    let numberOfItems = collectionView.numberOfItems(inSection: section)
+                    if numberOfItems > 0 {
+                        collectionView.deleteItems(
+                            at: (0..<numberOfItems).map({ IndexPath(item: $0, section: section) })
+                        )
+                    }
+                }
+            }
+        case (.loading, _):
```

Comment:
It doesn't handle `.loading` to `.loading` because that state change is handled above already, `.loading` to `.initial` can still be handled here correctly since the reason we want to handle loading → anything is to ensure we hide the visible spinner